<style>
 
{
  margin: 0;
  padding: 0;
  box-sizing: border-box;}

  body {
  margin-left: -24px;
  margin-top: 0;
  font-family: "Arial", sans-serif;
  width: 100%;
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  box-sizing: border-box;
}
  .title {
    text-align: center;
    font-size: 36px; 
    font-weight: bold;
    align-items: center;
    margin-top: 50px; 
  }
  .subtitle {
    text-align: center;
    font-size: 28px; 
    font-weight: bold;
    align-items: center;
    margin-bottom: 30px;
  }
  .logo {
    text-align: center;
    margin: 50px 0; 
    align-items: center;
  }
  .project-title {
    text-align: center;
    font-size: 28px; 
    font-weight: bold;
    align-items: center;
    margin: 40px 0; 
  }
  .authors {
    font-size: 20px; 
    text-align: center;
    margin-bottom: 50px;
    align-items: center;
  }
</style>

<div class="title">POLITECHNIKA LUBELSKA</div>
<div class="subtitle">WYDZIAŁ MATEMATYKI I INFORMATYKI TECHNICZNEJ</div>
<div class="subtitle">Kierunek: Inżynieria i Analiza Danych</div>

<div class="logo">
  <img src="logo/logopl.png" alt="Logo Politechniki Lubelskiej" width="300"> <!-- Większe logo -->
</div>

<div class="project-title">Projekt Zaliczeniowy z Zakresu Metod Eksploracji danych</div>
<div class="subtitle"><em>Praca wykonana przez:</em></div>
<div class="authors">
  <strong>Andrii Zapukhlyi, Nr albumu: s100935</strong> <br>
  <strong>Mateusz Drozd, Nr albumu: s100966</strong> <br>
</div>

## Zbiór Danych
Zbiór danych „Air Quality UCI” zawiera 9 358 instancji godzinnych uśrednionych odpowiedzi pięciu czujników półprzewodnikowych (metal-oxide) zainstalowanych w urządzeniu wieloczujnikowym do monitoringu jakości powietrza w terenie, wraz z jednoczesnymi pomiarami referencyjnymi stężeń zanieczyszczeń przez certyfikowany analizator.  
Dane obejmują okres od marca 2004 do lutego 2005 (13 miesięcy).  

| Nr | Nazwa zmiennej               | Opis                                                                                   |
| -- | ---------------------------- | -------------------------------------------------------------------------------------- |
| 0  | **Date**                     | Data pomiaru w formacie DD/MM/YYYY                                                     |
| 1  | **Time**                     | Godzina pomiaru w formacie HH.MM.SS                                                    |
| 2  | **True CO (mg/m³)**          | Rzeczywiste stężenie tlenku węgla (CO) zmierzone przez analizator referencyjny         |
| 3  | **PT08.S1 (tin oxide)**      | Odpowiedź sensora tlenku cyny (nominalnie celowanego w CO)                             |
| 4  | **True NMHC (µg/m³)**        | Rzeczywiste stężenie węglowodorów niemetanicznych (NMHC) przez analizator referencyjny |
| 5  | **True Benzene (µg/m³)**     | Rzeczywiste stężenie benzenu przez analizator referencyjny                             |
| 6  | **PT08.S2 (titania)**        | Odpowiedź sensora tlenku tytanu (nominalnie celowanego w NMHC)                         |
| 7  | **True NOx (ppb)**           | Rzeczywiste stężenie tlenków azotu (NOx) w ppb przez analizator referencyjny           |
| 8  | **PT08.S3 (tungsten oxide)** | Odpowiedź sensora tlenku wolframu (nominalnie celowanego w NOx)                        |
| 9  | **True NO₂ (µg/m³)**         | Rzeczywiste stężenie dwutlenku azotu (NO₂) przez analizator referencyjny               |
| 10 | **PT08.S4 (tungsten oxide)** | Odpowiedź sensora tlenku wolframu (nominalnie celowanego w NO₂)                        |
| 11 | **PT08.S5 (indium oxide)**   | Odpowiedź sensora tlenku indu (nominalnie celowanego w O₃)                             |
| 12 | **Temperature (°C)**         | Temperatura powietrza w stopniach Celsjusza                                            |
| 13 | **Relative Humidity (%)**    | Wilgotność względna powietrza w procentach                                             |
| 14 | **Absolute Humidity**        | Wilgotność bezwzględna (AH)                                                            |


### Charakterystyka danych
- **Typ:** Wielozmiennowy, szereg czasowy.  
- **Zadanie:** Regresja.  
- **Typ cech:** wartości rzeczywiste (float).  
- **Liczba instancji:** 9 358.  
- **Liczba cech:** 15.  

Brakujące wartości oznaczono wartością `-200` w komórkach sensorów i analizatorów.


## Cel projektu

Celem projektu jest zbudowanie modelu regresyjnego, który na podstawie dostępnych pomiarów środowiskowych będzie w stanie jak najdokładniej przewidzieć godzinową wartość stężenia benzenu (`C6H6(GT)`) w powietrzu. Benzen to toksyczny związek organiczny, uznawany za rakotwórczy i niebezpieczny dla zdrowia nawet w niskich stężeniach — dlatego jego skuteczne prognozowanie może mieć istotne znaczenie dla ochrony zdrowia publicznego, zarządzania jakością powietrza oraz systemów wczesnego ostrzegania.

Zmienna C6H6(GT) jest wartością ciągłą (wyrażoną w µg/m³), dlatego naturalnym podejściem analitycznym jest zastosowanie regresji nadzorowanej, której celem jest predykcja wartości liczbowej na podstawie zestawu cech wejściowych. Model może służyć m.in. do:

- przewidywania poziomu zanieczyszczenia w czasie rzeczywistym
- symulacji skutków zmian warunków atmosferycznych na jakość powietrza
- wspierania systemów ostrzegania w miastach i strefach przemysłowych

**W tym celu: wszystkie zmienne użyte do przewidywania (`C6H6(GT)`):**  
   - Odpowiedzi pięciu sensorów półprzewodnikowych (`PT08.S1(CO)`, `PT08.S2(NMHC)`, `PT08.S3(NOx)`, `PT08.S4(NO2)`, `PT08.S5(O3)`)  
   - Pomiary referencyjne innych gazów (`CO(GT)`, `NMHC(GT)`, `NOx(GT)`, `NO2(GT)`)  
   - Warunki meteorologiczne (`T`, `RH`, `AH`)  



In [1]:
import pandas as pd

df = pd.read_csv(
    'data\AirQualityUCI.csv',
    sep=';',
    decimal=',',
)
df.head().iloc[:,11:17]

PT08.S5(O3)     T    RH      AH  Unnamed: 15  Unnamed: 16
0       1268.0  13.6  48.9  0.7578          NaN          NaN
1        972.0  13.3  47.7  0.7255          NaN          NaN
2       1074.0  11.9  54.0  0.7502          NaN          NaN
3       1203.0  11.0  60.0  0.7867          NaN          NaN
4       1110.0  11.2  59.6  0.7888          NaN          NaN

Ponieważ kolumny `"Unnamed: 15"` i `"Unnamed: 16"` zawierają wyłącznie brakujące wartości (–200 lub NaN), usuwamy je przed dalszą analizą.

In [2]:
import numpy as np
df = df.drop(columns=["Unnamed: 15", "Unnamed: 16"])
df.replace(-200.0, np.nan, inplace=True)
df = df[:-114]

W ramach oczyszczania danych usunięto ostatnie 144 wiersze, które nie zawierały żadnych wartości (były w całości puste). Dzięki temu otrzymaliśmy spójniejszy zbiór, wolny od „pustych” rekordów, co ułatwia dalsze analizy statystyczne, wizualizacje oraz modelowanie.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         7674 non-null   float64
 3   PT08.S1(CO)    8991 non-null   float64
 4   NMHC(GT)       914 non-null    float64
 5   C6H6(GT)       8991 non-null   float64
 6   PT08.S2(NMHC)  8991 non-null   float64
 7   NOx(GT)        7718 non-null   float64
 8   PT08.S3(NOx)   8991 non-null   float64
 9   NO2(GT)        7715 non-null   float64
 10  PT08.S4(NO2)   8991 non-null   float64
 11  PT08.S5(O3)    8991 non-null   float64
 12  T              8991 non-null   float64
 13  RH             8991 non-null   float64
 14  AH             8991 non-null   float64
dtypes: float64(13), object(2)
memory usage: 1.1+ MB


## 1. Podstawowe wymiary i typy danych

* **Liczba wierszy (entry):** 9 357
* **Liczba kolumn:** 15
* **Typy danych:**

  * 2 kolumny obiektowe (`object`): `Date`, `Time`
  * 13 kolumn numerycznych (`float64`): pomiary gazów, sygnały czujników, T, RH, AH

---

## 2. Zestawienie braków danych

| Kolumna       | Non-Null | Braki (liczba) | Braki (%) |
| ------------- | -------: | -------------: | --------: |
| CO(GT)        |    7 674 |          1 683 |    18,0 % |
| PT08.S1(CO)   |    8 991 |            366 |     3,9 % |
| NMHC(GT)      |      914 |          8 443 |    90,2 % |
| C6H6(GT)      |    8 991 |            366 |     3,9 % |
| PT08.S2(NMHC) |    8 991 |            366 |     3,9 % |
| NOx(GT)       |    7 718 |          1 639 |    17,5 % |
| PT08.S3(NOx)  |    8 991 |            366 |     3,9 % |
| NO2(GT)       |    7 715 |          1 642 |    17,6 % |
| PT08.S4(NO2)  |    8 991 |            366 |     3,9 % |
| PT08.S5(O3)   |    8 991 |            366 |     3,9 % |
| T             |    8 991 |            366 |     3,9 % |
| RH            |    8 991 |            366 |     3,9 % |
| AH            |    8 991 |            366 |     3,9 % |

* **Najwięcej braków** w kolumnie `NMHC(GT)` (\~90 %).
* **Pozostałe kolumny** poza kilkoma wartościami w `CO(GT)`, `NOx(GT)`, `NO2(GT)` i oczywiście w samych datach/czasie mają ok. 4 % braków.




In [4]:
df.shape

(9357, 15)

Nasz zbior ma 9357 kolumn i 15 wierszy.

# Tabela statystyk opisowych dla każdej zmiennej

| Zmienna           |   Średnia |     Min |     Max | Odch. std. |   Wariancja | Skośność | Kurtoza |        Q1 |   Mediana |        Q3 |
| ----------------- | --------: | ------: | ------: | ---------: | ----------: | -------: | ------: | --------: | --------: | --------: |
| **PT08.S3(NOx)**  |  963.2975 |  461.00 | 1935.00 |   265.9142 |  70710.3448 |   0.8312 |  0.6530 |  769.0000 |  920.0000 | 1131.0000 |
| **NO2(GT)**       |  100.2600 |   19.00 |  196.00 |    31.4938 |    991.8609 |   0.0741 | –0.1430 |   78.5000 |   99.0000 |  122.0000 |
| **C6H6(GT)**      |   10.7711 |    0.50 |   39.20 |     7.4181 |     55.0287 |   0.9964 |  0.7950 |    4.8000 |    9.1000 |   14.8000 |
| **CO(GT)**        |    2.3536 |    0.30 |    8.10 |     1.4095 |      1.9867 |   1.0643 |  1.0700 |    1.3000 |    2.0000 |    3.1000 |
| **NMHC(GT)**      |  231.0254 |    7.00 | 1189.00 |   208.4619 |  43456.3686 |   1.4826 |  1.9560 |   77.0000 |  157.0000 |  318.5000 |
| **PT08.S2(NMHC)** |  966.1161 |  448.00 | 1754.00 |   266.4246 |  70982.0446 |   0.3504 | –0.4220 |  754.0000 |  944.0000 | 1142.5000 |
| **PT08.S1(CO)**   | 1207.8791 |  753.00 | 2040.00 |   241.8170 |  58475.4599 |   0.5644 | –0.1030 | 1017.0000 | 1172.0000 | 1380.0000 |
| **NOx(GT)**       |  143.5018 |   12.00 |  478.00 |    81.8297 |   6696.1026 |   0.8645 |  0.4560 |   81.0000 |  128.0000 |  187.0000 |
| **PT08.S4(NO2)**  | 1600.6203 |  955.00 | 2679.00 |   302.2918 |  91380.3278 |   0.6900 |  0.2730 | 1369.5000 | 1556.0000 | 1783.5000 |
| **PT08.S5(O3)**   | 1045.8126 |  263.00 | 2359.00 |   400.1347 | 160107.7481 |   0.3372 | –0.3510 |  760.0000 | 1009.0000 | 1320.0000 |
| **AH**            |    0.8319 |  0.4023 |  1.4852 |     0.1785 |      0.0319 |   0.7323 |  0.8370 |    0.7189 |    0.8177 |    0.9275 |
| **T**             |   15.6015 |  6.3000 | 30.0000 |     4.8253 |     23.2836 |   0.6252 |  0.0700 |   11.9000 |   15.0000 |   18.3000 |
| **RH**            |   49.0502 | 14.9000 | 83.2000 |    15.2667 |    233.0735 |   0.0081 | –0.8650 |   36.7000 |   49.6000 |   60.5500 |



## Wnioski z tabeli statystyk opisowych:

---

## 1. Zróżnicowanie skali i rozrzutu

* Największe wartości średnie i rozrzut mają **czujniki PT08.S1–PT08.S5** (zakresy sygnałów od \~250 do ponad 2 600 jednostek ADC). Ich odchylenia standardowe sięgają kilkuset, a wariancje dziesiątek tysięcy, co sugeruje wysoką zmienność w funkcji warunków atmosferycznych i stężeń.
* W porównaniu do nich **stężenia gazów** (CO, NO₂, NOₓ, C6H6, NMHC) mają znacznie mniejsze zakresy wartości oraz relatywnie mniejsze odchylenia standardowe.

---

## 2. Asymetria i „ogonki” rozkładów

* Najsilniej prawoskośne są:

  * **CO(GT)** (skośność ≈ 1,06)
  * **C6H6(GT)** (≈ 0,996)
  * **NMHC(GT)** (≈ 1,48)
    co oznacza, że od czasu do czasu pojawiają się bardzo wysokie pikowe wartości.
* Z kolei **wilgotność względna RH** ma prawie symetryczny rozkład (skośność ≈ 0,01), a **PT08.S2(NMHC)** i **PT08.S5(O₃)** są nawet lekko lewoskośne (skośność ujemna).

---

## 3. „Spiczastość” rozkładów (kurtoza)

* Najwyższą kurtozę (tzw. „spiczastość”) ma **NMHC(GT)** (≈ 1,96) – rozkład z ciężkimi ogonami i wyraźnym stożkowatym kształtem.
* Rozkład RH jest bardziej płaski niż normalny (kurtoza ≈ –0,865), co znaczy, że wartości skrajne trafiają się częściej niż w rozkładzie normalnym.

---

## 4. Kwantyle i mediany

* Dla większości zmiennych mediana (Q2) znajduje się znacząco poniżej średniej w przypadku prawoskośnych rozkładów, co wskazuje, że częściej spotykamy mniejsze wartości, a wysokie obserwacje podnoszą średnią.
* Zmienne bliskie symetrii (RH, NO2) mają medianę niemal równą średniej.

